### **Data Ingestion**

In [39]:
import numpy as np
import pandas as pd
from zipfile import ZipFile
from functools import reduce
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats
import os
from datetime import datetime, date, time

In [40]:
path = r'C:\Users\ncc\Downloads\olist_dataset.zip'
extracted_file = 'data_files'


In [41]:
with ZipFile(path, 'r') as f:
    f.extractall('data_files')

In [42]:
data_frames = {}

for file in os.listdir(extracted_file):
    if file.endswith('csv'):
        file_path = os.path.join(extracted_file, file)
        data_frames[file] = pd.read_csv(file_path)

In [43]:
for name, df in data_frames.items():
    print(f'{name}')

olist_customers_dataset.csv
olist_geolocation_dataset.csv
olist_orders_dataset.csv
olist_order_items_dataset.csv
olist_order_payments_dataset.csv
olist_order_reviews_dataset.csv
olist_products_dataset.csv
olist_sellers_dataset.csv
product_category_name_translation.csv


In [44]:
customers = pd.read_csv(r'data_files\olist_customers_dataset.csv')

print(customers.shape)
print(customers.columns)
customers.head(3)

(99441, 5)
Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state'],
      dtype='object')


customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP

In [45]:
geolocation = pd.read_csv(r'data_files\olist_geolocation_dataset.csv')
print(geolocation.shape)
print(geolocation.columns)
geolocation.head(3)

(1000163, 5)
Index(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng',
       'geolocation_city', 'geolocation_state'],
      dtype='object')


geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP

In [46]:
order_items = pd.read_csv(r'data_files\olist_order_items_dataset.csv')
print(order_items.shape)
print(order_items.columns)
order_items.head()

(112650, 7)
Index(['order_id', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value'],
      dtype='object')


order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [47]:
order_payment = pd.read_csv(r"data_files\olist_order_payments_dataset.csv")
print(order_payment.columns)
print(order_payment.shape)
order_payment.head()

Index(['order_id', 'payment_sequential', 'payment_type',
       'payment_installments', 'payment_value'],
      dtype='object')
(103886, 5)


order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [48]:
order_reviews = pd.read_csv(r"data_files\olist_order_reviews_dataset.csv")
print(order_reviews.columns)
print(order_reviews.shape)
order_reviews.head()

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='object')
(99224, 7)


review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [49]:
order_data = pd.read_csv(r"data_files\olist_orders_dataset.csv")
print(order_data.columns)
print(order_data.shape)
order_data.head()

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date'],
      dtype='object')
(99441, 8)


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [50]:
product = pd.read_csv(r"data_files\olist_products_dataset.csv")
print(product.columns)
print(product.shape)
product.head()

Index(['product_id', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm'],
      dtype='object')
(32951, 9)


product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [51]:
sellers = pd.read_csv(r'data_files\olist_sellers_dataset.csv')
print(order_reviews.columns)
print(order_reviews.shape)
sellers.head() 

Index(['review_id', 'order_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp'],
      dtype='object')
(99224, 7)


seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [52]:
product_category = pd.read_csv(r"data_files\product_category_name_translation.csv")
print(product_category.columns)
print(product_category.shape)
product_category.head()

Index(['product_category_name', 'product_category_name_english'], dtype='object')
(71, 2)


product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [53]:
for name, df in data_frames.items():
    print(f'{name}: {list(df.columns)}')

olist_customers_dataset.csv: ['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']
olist_geolocation_dataset.csv: ['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city', 'geolocation_state']
olist_orders_dataset.csv: ['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']
olist_order_items_dataset.csv: ['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']
olist_order_payments_dataset.csv: ['order_id', 'payment_sequential', 'payment_type', 'payment_installments', 'payment_value']
olist_order_reviews_dataset.csv: ['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']
olist_products_dataset.csv: ['product_id', 'product_category_name', 'pro

In [54]:
# Get sets of columns for each file
columns_per_file = {name: set(df.columns) for name, df in data_frames.items()}

# Get intersctioin( columns that appear in every file)
common_cols = reduce(set.intersection, columns_per_file.values())
print('Common columns across all CSVs:', common_cols)

Common columns across all CSVs: set()


**MERGING DATASET AND WHY THEY SHOULD BE MERGED**
1. **Customers ↔ Orders**

* **Key:** `customer_id`

* **Why:** Orders are made by customers. You can attach customer info (city, state, etc.) to each order.

Merge: `order_data` ↔ `customers`


2. **Orders ↔ Order Items**

* **Key:** `order_id`

* **Why:** Each order has one or more items. This links orders to products and sellers.

Merge: `order_data` ↔ `order_items`


3. **Order Items ↔ Products**

* **Key:** `product_id`

* **Why:** Each item corresponds to a product. You can bring in product details (category, dimensions, weight).

Merge: `order_items` ↔ `product`


4. **Products ↔ Product Category Translation**

* **Key:** `product_category_name`

* **Why:** Product categories are in Portuguese, and this dataset provides English translations.

Merge: `product` ↔ `product_category`


5. **Order Items ↔ Sellers**

* **Key:** `seller_id`

* **Why:** Each item is sold by a seller. This lets you connect sellers’ locations to orders.

Merge: `order_items` ↔ `sellers`


6. **Orders ↔ Payments**

* **Key:** `order_id`
* **Why:** Every order has a payment record (method, value, installments).

Merge: `order_data` ↔ `order_payment`


**Orders ↔ Reviews**

* **Key:** `order_id`

* **Why:** Customers leave reviews for orders (scores, comments)

Merge: `order_data` ↔ `order_reviews`


8. **Customers ↔ Geolocation**

* **Key:** `zip_code_prefix` (`customer_zip_code_prefix` vs `geolocation_zip_code_prefix`)

* **Why:** To attach latitude/longitude and city/state to customers.
Note: Multiple geolocations can share the same ZIP prefix → requires aggregation or nearest match.

Merge: `customers` ↔ `geolocation`



In [55]:
#1️. Customers + Orders

orders_customers = pd.merge(order_data, customers, on='customer_id', how='inner')
orders_customers.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
2           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   
3           2017-12-15 00:00:00  7c142cf63193a1473d2e66489a9ae977   
4           2018-02-26 00:00:00  72632f0f9dd73dfee390c9b22eb56dd6   

   customer_zip_code_prefix            customer_city customer_state  
0                      3149                sao paulo             SP  
1                     47813                barreiras             BA  
2                     75265               vianopolis             GO  
3                     59296  sao goncalo do amarante             RN  
4                      9195              santo andre             SP

In [56]:
#2. Customers + Orders
orders_payments = pd.merge(orders_customers, order_payment, on='order_id', how='left')
orders_payments.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
4           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix customer_city customer_state  payment_sequential  \
0                      3149     sao paulo             SP                 1.0   
1                      3149     sao paulo             SP                 3.0   
2                      3149     sao paulo             SP                 2.0   
3                     47813     barreiras             BA                 1.0   
4                     75265    vianopolis             GO                 1.0   

  payment_type  payment_installments  payment_value  
0  credit_card                   1.0          18.12  
1      voucher                   1.0           2.00  
2      voucher                   1.0          18.59  
3       boleto                   1.0         141.46  
4  credit_card                   3.0         179.12

In [57]:
# 3. Orders + Reviews

orders_reviews = pd.merge(orders_payments, order_reviews, on='order_id', how='right')
orders_reviews.head()

order_id                       customer_id  \
0  73fc7af87114b39712e6da79b0a377eb  41dcb106f807e993532d446263290104   
1  a548910a1c6147796b98fdf73dbeba33  8a2e7ef9053dea531e4dc76bd6d853e6   
2  f9e4b658b201a9f2ecdecbb34bed034b  e226dfed6544df5b7b87a48208690feb   
3  658677c97b385a9be170737859d3511b  de6dff97e5f1ba84a3cd9a3bc97df5f6   
4  8e6bfb81e283fa7e4f11123a3fb894f1  5986b333ca0d44534a156a52a8e33a83   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2018-01-11 15:30:49  2018-01-11 15:47:59   
1    delivered      2018-02-28 12:25:19  2018-02-28 12:48:39   
2    delivered      2018-02-03 09:56:22  2018-02-03 10:33:41   
3    delivered      2017-04-09 17:41:13  2017-04-09 17:55:19   
4    delivered      2018-02-10 10:59:03  2018-02-10 15:48:21   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2018-01-12 21:57:22           2018-01-17 18:42:41   
1          2018-03-02 19:08:15           2018-03-09 23:17:20   
2          2018-02-06 16:18:28           2018-02-16 17:28:48   
3          2017-04-10 14:24:47           2017-04-20 09:08:35   
4          2018-02-15 19:36:14           2018-02-28 16:33:35   

  order_estimated_delivery_date                customer_unique_id  \
0           2018-02-02 00:00:00  68a5590b9926689be4e10f4ae2db21a8   
1           2018-03-14 00:00:00  64190b91b656ab8f37eb89b93dc84584   
2           2018-03-09 00:00:00  1d47144362c14e94ccdd213e8ec277d5   
3           2017-05-10 00:00:00  c8cf6cb6b838dc7a33ed199b825e8616   
4           2018-03-09 00:00:00  d16000272660a1fef81482ad75ba572a   

   customer_zip_code_prefix  ... payment_sequential payment_type  \
0                      6030  ...                1.0  credit_card   
1                     13380  ...                1.0  credit_card   
2                     44571  ...                1.0  credit_card   
3                     88735  ...                1.0  credit_card   
4                     89520  ...                1.0  credit_card   

   payment_installments payment_value                         review_id  \
0                   8.0        397.26  7bc2406110b926393aa56f80a40eba40   
1                   1.0         88.09  80e641a11e56f04c1ad469d5645fdfde   
2                   1.0        194.12  228ce5500dc1d8e020d8d1322874b6f0   
3                   1.0        222.84  e64fb393e7b32834bb789ff8bb30750e   
4                  10.0       1333.25  f7c4243c7fe1938f181bec41a392bdeb   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53  

[5 rows x 22 columns]

In [58]:
# 4. Orders + Order Items
orders_items = pd.merge(orders_reviews, order_items, on='order_id', how='outer')
orders_items.head()

order_id                       customer_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  3ce436f183e68e07877b285a838db11a   
1  00018f77f2f0320c557190d7a144bdd3  f6dd3ec061db4e3987629fe6b26e5cce   
2  000229ec398224ef6ca0657da4fc703e  6489ae5e4333f3693df5ad4372dab6d3   
3  00024acbcdf0a6daa1e931b038114c75  d4eb9395c8c0431ee92fce09860c5a06   
4  00042b26cf59d7ce69dfabb4e55b4fd9  58dbd0b2d70206bf40e62cd34e84d795   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-09-13 08:59:02  2017-09-13 09:45:35   
1    delivered      2017-04-26 10:53:06  2017-04-26 11:05:13   
2    delivered      2018-01-14 14:33:31  2018-01-14 14:48:30   
3    delivered      2018-08-08 10:00:35  2018-08-08 10:10:18   
4    delivered      2017-02-04 13:57:51  2017-02-04 14:10:13   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-09-19 18:34:16           2017-09-20 23:43:48   
1          2017-05-04 14:35:00           2017-05-12 16:04:24   
2          2018-01-16 12:36:48           2018-01-22 13:19:16   
3          2018-08-10 13:28:00           2018-08-14 13:32:39   
4          2017-02-16 09:46:09           2017-03-01 16:42:31   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-09-29 00:00:00  871766c5855e863f6eccc05f988b23cb   
1           2017-05-15 00:00:00  eb28e67c4c0b83846050ddfb8a35d051   
2           2018-02-05 00:00:00  3818d81c6709e39d06b2738a8d3a2474   
3           2018-08-20 00:00:00  af861d436cfc08b2c2ddefd0ba074622   
4           2017-03-17 00:00:00  64b576fb70d441e8f1b2d7d446e483c5   

   customer_zip_code_prefix  ... review_comment_title  \
0                   28013.0  ...                  NaN   
1                   15775.0  ...                  NaN   
2                   35661.0  ...                  NaN   
3                   12952.0  ...                  NaN   
4                   13226.0  ...                  NaN   

                              review_comment_message  review_creation_date  \
0     Perfeito, produto entregue antes do combinado.   2017-09-21 00:00:00   
1                                                NaN   2017-05-13 00:00:00   
2  Chegou antes do prazo previsto e o produto sur...   2018-01-23 00:00:00   
3                                                NaN   2018-08-15 00:00:00   
4            Gostei pois veio no prazo determinado .   2017-03-02 00:00:00   

  review_answer_timestamp  order_item_id                        product_id  \
0     2017-09-22 10:57:03            1.0  4244733e06e7ecb4970a6e2683c13e61   
1     2017-05-15 11:34:13            1.0  e5f2d52b802189ee658865ca93d83a8f   
2     2018-01-23 16:06:31            1.0  c777355d18b72b67abbeef9df44fd0fd   
3     2018-08-15 16:39:01            1.0  7634da152a4610f1595efa32f14722fc   
4     2017-03-03 10:54:59            1.0  ac6c3623068f30de03045865e4e10089   

                          seller_id  shipping_limit_date   price freight_value  
0  48436dade18ac8b2bce089ec2a041202  2017-09-19 09:45:35   58.90         13.29  
1  dd7ddc04e1b6c2c614352b383efe2d36  2017-05-03 11:05:13  239.90         19.93  
2  5b51032eddd242adc84c38acab88f23d  2018-01-18 14:48:30  199.00         17.87  
3  9d7a1d34a5052409006425275ba1c2b4  2018-08-15 10:10:18   12.99         12.79  
4  df560393f3a51e74553ab94004ba5c87  2017-02-13 13:57:51  199.90         18.14  

[5 rows x 28 columns]

In [59]:
# 5. Order Items + Products
orders_products = pd.merge(orders_items, product, on='product_id', how='left')
orders_products.head()

order_id                       customer_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  3ce436f183e68e07877b285a838db11a   
1  00018f77f2f0320c557190d7a144bdd3  f6dd3ec061db4e3987629fe6b26e5cce   
2  000229ec398224ef6ca0657da4fc703e  6489ae5e4333f3693df5ad4372dab6d3   
3  00024acbcdf0a6daa1e931b038114c75  d4eb9395c8c0431ee92fce09860c5a06   
4  00042b26cf59d7ce69dfabb4e55b4fd9  58dbd0b2d70206bf40e62cd34e84d795   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-09-13 08:59:02  2017-09-13 09:45:35   
1    delivered      2017-04-26 10:53:06  2017-04-26 11:05:13   
2    delivered      2018-01-14 14:33:31  2018-01-14 14:48:30   
3    delivered      2018-08-08 10:00:35  2018-08-08 10:10:18   
4    delivered      2017-02-04 13:57:51  2017-02-04 14:10:13   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-09-19 18:34:16           2017-09-20 23:43:48   
1          2017-05-04 14:35:00           2017-05-12 16:04:24   
2          2018-01-16 12:36:48           2018-01-22 13:19:16   
3          2018-08-10 13:28:00           2018-08-14 13:32:39   
4          2017-02-16 09:46:09           2017-03-01 16:42:31   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-09-29 00:00:00  871766c5855e863f6eccc05f988b23cb   
1           2017-05-15 00:00:00  eb28e67c4c0b83846050ddfb8a35d051   
2           2018-02-05 00:00:00  3818d81c6709e39d06b2738a8d3a2474   
3           2018-08-20 00:00:00  af861d436cfc08b2c2ddefd0ba074622   
4           2017-03-17 00:00:00  64b576fb70d441e8f1b2d7d446e483c5   

   customer_zip_code_prefix  ...   price freight_value  product_category_name  \
0                   28013.0  ...   58.90         13.29             cool_stuff   
1                   15775.0  ...  239.90         19.93               pet_shop   
2                   35661.0  ...  199.00         17.87       moveis_decoracao   
3                   12952.0  ...   12.99         12.79             perfumaria   
4                   13226.0  ...  199.90         18.14     ferramentas_jardim   

  product_name_lenght  product_description_lenght  product_photos_qty  \
0                58.0                       598.0                 4.0   
1                56.0                       239.0                 2.0   
2                59.0                       695.0                 2.0   
3                42.0                       480.0                 1.0   
4                59.0                       409.0                 1.0   

  product_weight_g  product_length_cm product_height_cm product_width_cm  
0            650.0               28.0               9.0             14.0  
1          30000.0               50.0              30.0             40.0  
2           3050.0               33.0              13.0             33.0  
3            200.0               16.0              10.0             15.0  
4           3750.0               35.0              40.0             30.0  

[5 rows x 36 columns]

In [60]:
# 6. Products + Products Category Translation
orders_products = pd.merge(orders_products, product_category, on='product_category_name', how='left')
orders_products.head()

order_id                       customer_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  3ce436f183e68e07877b285a838db11a   
1  00018f77f2f0320c557190d7a144bdd3  f6dd3ec061db4e3987629fe6b26e5cce   
2  000229ec398224ef6ca0657da4fc703e  6489ae5e4333f3693df5ad4372dab6d3   
3  00024acbcdf0a6daa1e931b038114c75  d4eb9395c8c0431ee92fce09860c5a06   
4  00042b26cf59d7ce69dfabb4e55b4fd9  58dbd0b2d70206bf40e62cd34e84d795   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-09-13 08:59:02  2017-09-13 09:45:35   
1    delivered      2017-04-26 10:53:06  2017-04-26 11:05:13   
2    delivered      2018-01-14 14:33:31  2018-01-14 14:48:30   
3    delivered      2018-08-08 10:00:35  2018-08-08 10:10:18   
4    delivered      2017-02-04 13:57:51  2017-02-04 14:10:13   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-09-19 18:34:16           2017-09-20 23:43:48   
1          2017-05-04 14:35:00           2017-05-12 16:04:24   
2          2018-01-16 12:36:48           2018-01-22 13:19:16   
3          2018-08-10 13:28:00           2018-08-14 13:32:39   
4          2017-02-16 09:46:09           2017-03-01 16:42:31   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-09-29 00:00:00  871766c5855e863f6eccc05f988b23cb   
1           2017-05-15 00:00:00  eb28e67c4c0b83846050ddfb8a35d051   
2           2018-02-05 00:00:00  3818d81c6709e39d06b2738a8d3a2474   
3           2018-08-20 00:00:00  af861d436cfc08b2c2ddefd0ba074622   
4           2017-03-17 00:00:00  64b576fb70d441e8f1b2d7d446e483c5   

   customer_zip_code_prefix  ... freight_value product_category_name  \
0                   28013.0  ...         13.29            cool_stuff   
1                   15775.0  ...         19.93              pet_shop   
2                   35661.0  ...         17.87      moveis_decoracao   
3                   12952.0  ...         12.79            perfumaria   
4                   13226.0  ...         18.14    ferramentas_jardim   

   product_name_lenght product_description_lenght  product_photos_qty  \
0                 58.0                      598.0                 4.0   
1                 56.0                      239.0                 2.0   
2                 59.0                      695.0                 2.0   
3                 42.0                      480.0                 1.0   
4                 59.0                      409.0                 1.0   

   product_weight_g product_length_cm  product_height_cm product_width_cm  \
0             650.0              28.0                9.0             14.0   
1           30000.0              50.0               30.0             40.0   
2            3050.0              33.0               13.0             33.0   
3             200.0              16.0               10.0             15.0   
4            3750.0              35.0               40.0             30.0   

  product_category_name_english  
0                    cool_stuff  
1                      pet_shop  
2               furniture_decor  
3                     perfumery  
4                  garden_tools  

[5 rows x 37 columns]

In [61]:
# Order Items + Sellers

orders_sellers = pd.merge(orders_products, sellers, on='seller_id', how='inner')
orders_sellers.head()

order_id                       customer_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  3ce436f183e68e07877b285a838db11a   
1  00018f77f2f0320c557190d7a144bdd3  f6dd3ec061db4e3987629fe6b26e5cce   
2  000229ec398224ef6ca0657da4fc703e  6489ae5e4333f3693df5ad4372dab6d3   
3  00024acbcdf0a6daa1e931b038114c75  d4eb9395c8c0431ee92fce09860c5a06   
4  00042b26cf59d7ce69dfabb4e55b4fd9  58dbd0b2d70206bf40e62cd34e84d795   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-09-13 08:59:02  2017-09-13 09:45:35   
1    delivered      2017-04-26 10:53:06  2017-04-26 11:05:13   
2    delivered      2018-01-14 14:33:31  2018-01-14 14:48:30   
3    delivered      2018-08-08 10:00:35  2018-08-08 10:10:18   
4    delivered      2017-02-04 13:57:51  2017-02-04 14:10:13   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-09-19 18:34:16           2017-09-20 23:43:48   
1          2017-05-04 14:35:00           2017-05-12 16:04:24   
2          2018-01-16 12:36:48           2018-01-22 13:19:16   
3          2018-08-10 13:28:00           2018-08-14 13:32:39   
4          2017-02-16 09:46:09           2017-03-01 16:42:31   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-09-29 00:00:00  871766c5855e863f6eccc05f988b23cb   
1           2017-05-15 00:00:00  eb28e67c4c0b83846050ddfb8a35d051   
2           2018-02-05 00:00:00  3818d81c6709e39d06b2738a8d3a2474   
3           2018-08-20 00:00:00  af861d436cfc08b2c2ddefd0ba074622   
4           2017-03-17 00:00:00  64b576fb70d441e8f1b2d7d446e483c5   

   customer_zip_code_prefix  ... product_description_lenght  \
0                   28013.0  ...                      598.0   
1                   15775.0  ...                      239.0   
2                   35661.0  ...                      695.0   
3                   12952.0  ...                      480.0   
4                   13226.0  ...                      409.0   

  product_photos_qty  product_weight_g product_length_cm  product_height_cm  \
0                4.0             650.0              28.0                9.0   
1                2.0           30000.0              50.0               30.0   
2                2.0            3050.0              33.0               13.0   
3                1.0             200.0              16.0               10.0   
4                1.0            3750.0              35.0               40.0   

   product_width_cm product_category_name_english  seller_zip_code_prefix  \
0              14.0                    cool_stuff                   27277   
1              40.0                      pet_shop                    3471   
2              33.0               furniture_decor                   37564   
3              15.0                     perfumery                   14403   
4              30.0                  garden_tools                   87900   

     seller_city seller_state  
0  volta redonda           SP  
1      sao paulo           SP  
2  borda da mata           MG  
3         franca           SP  
4         loanda           PR  

[5 rows x 40 columns]

In [62]:
# 8. Customers + Geolocation (via zip prefix)

#Note that there are many geolocations per ZIP, so we first occurence per prefix
geo_unique = geolocation.groupby('geolocation_zip_code_prefix').first().reset_index()
geo_unique.head(2)
# final_df = pd.merge(orders_sellers, geo_unique, left_on='customer_zip_code_prefix', right_on= 'geolocation_zip_code_prefix', how='left')

# print(final_df.shape)
# final_df.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1001       -23.549292       -46.633559   
1                         1002       -23.548318       -46.635421   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP

In [63]:
# 8. Customers + Geolocation (via zip prefix)

#Note that there are many geolocations per ZIP, so we first occurence per prefix
geo_unique = geolocation.groupby('geolocation_zip_code_prefix').first().reset_index()

final_df = pd.merge(orders_sellers, geo_unique, left_on='customer_zip_code_prefix', right_on= 'geolocation_zip_code_prefix', how='left')

print(final_df.shape)
final_df.head()

(118274, 45)


order_id                       customer_id  \
0  00010242fe8c5a6d1ba2dd792cb16214  3ce436f183e68e07877b285a838db11a   
1  00018f77f2f0320c557190d7a144bdd3  f6dd3ec061db4e3987629fe6b26e5cce   
2  000229ec398224ef6ca0657da4fc703e  6489ae5e4333f3693df5ad4372dab6d3   
3  00024acbcdf0a6daa1e931b038114c75  d4eb9395c8c0431ee92fce09860c5a06   
4  00042b26cf59d7ce69dfabb4e55b4fd9  58dbd0b2d70206bf40e62cd34e84d795   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-09-13 08:59:02  2017-09-13 09:45:35   
1    delivered      2017-04-26 10:53:06  2017-04-26 11:05:13   
2    delivered      2018-01-14 14:33:31  2018-01-14 14:48:30   
3    delivered      2018-08-08 10:00:35  2018-08-08 10:10:18   
4    delivered      2017-02-04 13:57:51  2017-02-04 14:10:13   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-09-19 18:34:16           2017-09-20 23:43:48   
1          2017-05-04 14:35:00           2017-05-12 16:04:24   
2          2018-01-16 12:36:48           2018-01-22 13:19:16   
3          2018-08-10 13:28:00           2018-08-14 13:32:39   
4          2017-02-16 09:46:09           2017-03-01 16:42:31   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-09-29 00:00:00  871766c5855e863f6eccc05f988b23cb   
1           2017-05-15 00:00:00  eb28e67c4c0b83846050ddfb8a35d051   
2           2018-02-05 00:00:00  3818d81c6709e39d06b2738a8d3a2474   
3           2018-08-20 00:00:00  af861d436cfc08b2c2ddefd0ba074622   
4           2017-03-17 00:00:00  64b576fb70d441e8f1b2d7d446e483c5   

   customer_zip_code_prefix  ... product_width_cm  \
0                   28013.0  ...             14.0   
1                   15775.0  ...             40.0   
2                   35661.0  ...             33.0   
3                   12952.0  ...             15.0   
4                   13226.0  ...             30.0   

  product_category_name_english  seller_zip_code_prefix    seller_city  \
0                    cool_stuff                   27277  volta redonda   
1                      pet_shop                    3471      sao paulo   
2               furniture_decor                   37564  borda da mata   
3                     perfumery                   14403         franca   
4                  garden_tools                   87900         loanda   

   seller_state  geolocation_zip_code_prefix geolocation_lat  geolocation_lng  \
0            SP                      28013.0      -21.758076       -41.312633   
1            SP                      15775.0      -20.212393       -50.941471   
2            MG                      35661.0      -19.860439       -44.597972   
3            SP                      12952.0      -23.144923       -46.539830   
4            PR                      13226.0      -23.249008       -46.824961   

        geolocation_city geolocation_state  
0  campos dos goytacazes                RJ  
1        santa fe do sul                SP  
2          pará de minas                MG  
3                atibaia                SP  
4        varzea paulista                SP  

[5 rows x 45 columns]

#### **PROBABILITY**

______________________________________________________________________
______________________________________________________________________
1. Probability of an Event

Concept: Basic probability = favorable outcomes / total outcomes.

Question: What is the probability that a randomly chosen order received a 5-star review?

In [64]:
#We will be using the order_review data to answer this queation
orders_reviews.head(2)

order_id                       customer_id  \
0  73fc7af87114b39712e6da79b0a377eb  41dcb106f807e993532d446263290104   
1  a548910a1c6147796b98fdf73dbeba33  8a2e7ef9053dea531e4dc76bd6d853e6   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2018-01-11 15:30:49  2018-01-11 15:47:59   
1    delivered      2018-02-28 12:25:19  2018-02-28 12:48:39   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2018-01-12 21:57:22           2018-01-17 18:42:41   
1          2018-03-02 19:08:15           2018-03-09 23:17:20   

  order_estimated_delivery_date                customer_unique_id  \
0           2018-02-02 00:00:00  68a5590b9926689be4e10f4ae2db21a8   
1           2018-03-14 00:00:00  64190b91b656ab8f37eb89b93dc84584   

   customer_zip_code_prefix  ... payment_sequential payment_type  \
0                      6030  ...                1.0  credit_card   
1                     13380  ...                1.0  credit_card   

   payment_installments payment_value                         review_id  \
0                   8.0        397.26  7bc2406110b926393aa56f80a40eba40   
1                   1.0         88.09  80e641a11e56f04c1ad469d5645fdfde   

   review_score review_comment_title  review_comment_message  \
0             4                  NaN                     NaN   
1             5                  NaN                     NaN   

  review_creation_date review_answer_timestamp  
0  2018-01-18 00:00:00     2018-01-18 21:46:59  
1  2018-03-10 00:00:00     2018-03-11 03:05:13  

[2 rows x 22 columns]

In [65]:
# Probability of 5-star review
# Out of all orders, what is the probability that a review has a 5-star rating?
# This tells us how likely a customer is to give the maximum satisfaction score.

# p(5star) = number of 5-stars review/total reviews

total_reviews = len(order_reviews)
favourable = (order_reviews['review_score'] == 5).sum()

p_5star = round((favourable/total_reviews) * 100, 2)

print(f'Probability of 5 stars = {p_5star}')

Probability of 5 stars = 57.78


**Interpretation of result**

Approximately 58% of customers give 5 stars. This is the baseline "success" probability of excellent reviews.
_______________________________________________________________
_______________________________________________________________

2. Complementary Probability

Concept: 
𝑃(A complement) = 1−P(A)

Question: What is the probability that a review is not 5 stars?

In [66]:
#If we know the probability of 5 stars, what’s the probability of not getting a 5-star review?
#This means any rating from 1 to 4.

# We are still going to use the review dataset
# Formula style

p_not5 = ((((order_reviews['review_score'] != 5).sum())/ total_reviews) * 100)

print(f'Probability that reviews != 5 stars = {round(p_not5, 2)}%')

Probability that reviews != 5 stars = 42.22%


Interpretation
Approximately 42% of reviews are less than perfect. This highlights that not everyone(about half the customers) is satisfied with the product purchased. 

3. Joint Probability

Concept: P(A∩B)

Question: What is the probability that an order was paid by credit card and received a 5-star review?

In [67]:
#This gives the likelihood of two events happening together.

# Here we will be using the orders_payments and orders_reviews. And we have merged them already
# lets merge orders_payments and order_reviews

order_reviews.head()

orders_payment_reviews = pd.merge(orders_reviews, orders_payments, on='order_id', how='right')
orders_payment_reviews.head(3)

order_id                     customer_id_x  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status_x order_purchase_timestamp_x  order_approved_at_x  \
0      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
1      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
2      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date_x order_delivered_customer_date_x  \
0            2017-10-04 19:55:00             2017-10-10 21:25:13   
1            2017-10-04 19:55:00             2017-10-10 21:25:13   
2            2017-10-04 19:55:00             2017-10-10 21:25:13   

  order_estimated_delivery_date_x              customer_unique_id_x  \
0             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix_x  ... order_delivered_customer_date_y  \
0                      3149.0  ...             2017-10-10 21:25:13   
1                      3149.0  ...             2017-10-10 21:25:13   
2                      3149.0  ...             2017-10-10 21:25:13   

  order_estimated_delivery_date_y              customer_unique_id_y  \
0             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

  customer_zip_code_prefix_y  customer_city_y  customer_state_y  \
0                       3149        sao paulo                SP   
1                       3149        sao paulo                SP   
2                       3149        sao paulo                SP   

  payment_sequential_y  payment_type_y payment_installments_y payment_value_y  
0                  1.0     credit_card                    1.0           18.12  
1                  1.0     credit_card                    1.0           18.12  
2                  1.0     credit_card                    1.0           18.12  

[3 rows x 37 columns]

In [68]:
orders_payment_reviews.columns

Index(['order_id', 'customer_id_x', 'order_status_x',
       'order_purchase_timestamp_x', 'order_approved_at_x',
       'order_delivered_carrier_date_x', 'order_delivered_customer_date_x',
       'order_estimated_delivery_date_x', 'customer_unique_id_x',
       'customer_zip_code_prefix_x', 'customer_city_x', 'customer_state_x',
       'payment_sequential_x', 'payment_type_x', 'payment_installments_x',
       'payment_value_x', 'review_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp', 'customer_id_y', 'order_status_y',
       'order_purchase_timestamp_y', 'order_approved_at_y',
       'order_delivered_carrier_date_y', 'order_delivered_customer_date_y',
       'order_estimated_delivery_date_y', 'customer_unique_id_y',
       'customer_zip_code_prefix_y', 'customer_city_y', 'customer_state_y',
       'payment_sequential_y', 'payment_type_y', 'payment_installments_y',
       'payment_value_y'],
      dt

print(f'$P(Credit Card \!= 5 stars)$')

 $$ A \cap B $$

In [69]:
total_orders = len(orders_payment_reviews)
favourable_joint = ((orders_payment_reviews['payment_type_y'] == 'credit_card') & (orders_payment_reviews['review_score'] == 5)).sum()

p_joint = favourable_joint / total_orders

print(f"P(Credit Card ∩ 5 stars) = {round(p_joint *100, 2)}%")

P(Credit Card ∩ 5 stars) = 37.54%


**Interpretation**

Approximately 38% of all orders are both paid by credit card and get a perfect review.

__________________________________________________________________________
___________________________________________________________________________

4. Conditional Probability

Concept:

`𝑃(𝐴∣𝐵)=𝑃(𝐴∩𝐵)/𝑃(𝐵)` 
                
`𝑃(𝐴∣𝐵)= Favourable_outcomes / Total_credit_card_outcomes`



Question: What is the probability that a review is 5 stars given that the payment was made by credit card?

In [70]:
# If we already know the payment method was credit card, what is the probability that the review is 5 stars?
# This measures satisfaction among credit card users only.
# We are solving for independent conditional probability

# We might took few steps, and solve beyond the question for proper understanding
# That is, P(5 stars∣payment method = X)

#𝑃(5∣credit_card)
#P(5∣voucher)
#P(5∣boleto)
# Each is a separate conditional distribution. They are not mutually exclusive outcomes of the same experiment.

favourable_cond = ((orders_payment_reviews['payment_type_y'] == 'credit_card') & (orders_payment_reviews['review_score'] == 5)).sum()

total_credit = (orders_payment_reviews['payment_type_y'] == 'credit_card').sum()

p_cond = favourable_cond / total_credit

print(f'P(5 stars | Credit Card) = {round(p_cond*100, 2)}%')


favourable_cond = ((orders_payment_reviews['payment_type_y'] == 'voucher') &  (orders_payment_reviews["review_score"]==5)).sum()

total_credit = (orders_payment_reviews["payment_type_y"]=="voucher").sum()

p_cond = favourable_cond / total_credit

print(f"P(5 stars | voucher) = {round(p_cond *100, 2)}%")

#output: P(5 stars | voucher) = 55.85%


favourable_cond = ((orders_payment_reviews["payment_type_y"]=="boleto") &
                   (orders_payment_reviews["review_score"]==5)).sum()

total_credit = (orders_payment_reviews["payment_type_y"]=="boleto").sum()

p_cond = favourable_cond / total_credit
print(f"P(5 stars | boleto) = {round(p_cond *100, 2)}%")

#output: P(5 stars | boleto) = 56.95%


P(5 stars | Credit Card) = 57.41%
P(5 stars | voucher) = 55.85%
P(5 stars | boleto) = 56.95%


**Interpretation**

Each one is around 55–57%, meaning in every payment method, about half of customers gave 5 stars.

_______________________________________________________________________________
_______________________________________________________________________________

5. Bayes’ Rule 

Concept:

`P(A∣B )= P(B∣A)⋅P(A) / P(B)`

A = payment

B = review =5

`P(payment | review=5) = P(review=5 | payment).p(payment) / p(review =5)`


Question: Given that an order received a 5-star review, what is the probability that it was paid with each payment method (credit_card, boleto, voucher)?


Among only the 5-star reviews, how are payment methods distributed?

In [ ]:
# For credit card payment;  A = Credit card; B = Review(5). B remains constant for all payment types

# To get the probability of A
A_credit = (orders_payment_reviews['payment_type_y'] == "credit_card").sum()
total_payment = len(orders_payment_reviews['payment_type_y'])
p_credit = round((A_credit / total_payment)*100, 4)
print(f'P(A) = {p_credit}')

# To get the probability of B
B = (orders_payment_reviews["review_score"]==5).sum()
total_review = len(orders_payment_reviews['review_score'])
p_B = round((B / total_review)*100, 4)
print(f'P(B) = {p_B}')

# To get P(review=5 | payment)
# Number of B | A
favourable_cond = ((orders_payment_reviews["review_score"]==5) & 
                   (orders_payment_reviews["payment_type_y"]=="credit_card")).sum()

# Total number of 5-star reviews (denominator for all)
total_5star = (orders_payment_reviews["review_score"] == 5).sum()
p_BA = favourable_cond / total_5star

final_probability = (p_BA * p_credit) / p_B
print(f"P(Credit Card | 5 stars) = {round(final_probability * 100, 2)}%")




# For voucher payment, A = Voucher; B = Review(5)
# To get the probability of A
A_voucher = (orders_payment_reviews['payment_type_y'] == "voucher").sum()
total_payment = len(orders_payment_reviews['payment_type_y'])
p_voucher = round((A_voucher / total_payment)*100, 2)
print(f'\nP(A) = {p_voucher}')

# To get P(review=5 | payment)
# Number of B | A
favourable_cond = ((orders_payment_reviews["payment_type_y"]=="voucher") & 
                   (orders_payment_reviews["review_score"]==5)).sum()

# Total number of 5-star reviews (denominator for all)
total_5star = (orders_payment_reviews["review_score"] == 5).sum()
p_BA = favourable_cond / total_5star
final_probability = (p_BA * p_voucher) / p_B
print(f"P(Voucher | 5 stars) = {round(final_probability * 100, 2)}%")



# For boleto payment, A = Credit card; B = Review(5)
# To get the probability of A
A_boleto = (orders_payment_reviews['payment_type_y'] == "boleto").sum()
total_payment = len(orders_payment_reviews['payment_type_y'])
p_boleto = round((A_boleto / total_payment)*100, 2)
print(f'\nP(A) = {p_boleto}')

# To get P(review=5 | payment)
# Number of B | A
favourable_cond = ((orders_payment_reviews["payment_type_y"]=="boleto") & 
                   (orders_payment_reviews["review_score"]==5)).sum()

# Total number of 5-star reviews (denominator for all)
total_5star = (orders_payment_reviews["review_score"] == 5).sum()
p_BA = favourable_cond / total_5star

final_probability = (p_BA * p_boleto) / p_B
print(f"P(Boleto| 5 stars) = {round(final_probability * 100, 2)}%")


# For debit card payment
# To get the probability of A
A_debit = (orders_payment_reviews['payment_type_y'] == "debit_card").sum()
total_payment = len(orders_payment_reviews['payment_type_y'])
p_debit = round((A_debit / total_payment)*100, 2)
print(f'P(A) = {p_debit}')

# To get P(review=5 | payment)
# Number of B | A
favourable_cond = ((orders_payment_reviews["payment_type_y"]=="debit_card") & 
                   (orders_payment_reviews["review_score"]==5)).sum()

# Total number of 5-star reviews (denominator for all)
total_5star = (orders_payment_reviews["review_score"] == 5).sum()
p_BA = favourable_cond / total_5star
final_probability = (p_BA * p_debit) / p_B
print(f"P(Debit Card | 5 stars) = {round(final_probability * 100, 2)}%")


# To get for the undefined payment type
A_undefined = (orders_payment_reviews['payment_type_y'] == "not_defined").sum()
total_payment = len(orders_payment_reviews['payment_type_y'])
p_undefined = round((A_undefined / total_payment)*100, 2)
print(f'\nP(A) = {p_undefined}')

# To get P(review=5 | payment)
# Number of B | A
favourable_cond = ((orders_payment_reviews["payment_type_y"]=="not_defined") & 
                   (orders_payment_reviews["review_score"]==5)).sum()

# Total number of 5-star reviews (denominator for all)
total_5star = (orders_payment_reviews["review_score"] == 5).sum()
p_BA = favourable_cond / total_5star
final_probability = (p_BA * p_undefined) / p_B
print(f"\nP(Undefined Payment | 5 stars) = {round(final_probability * 100, 2)}%")

P(A) = 65.39
P(B) = 57.1
P(Credit Card | 5 stars) = 42.99%

P(A) = 17.27
P(Voucher | 5 stars) = 2.92%

P(A) = 16.09

P(Boleto | 5 stars) = 2.58%
P(A) = 1.24
P(Debit Card | 5 stars) = 0.02%

P(A) = 0.0
P(Undefined | 5 stars) = 0.0%


In [94]:
# For credit card payment;  A = Credit card; B = Review(5). B remains constant for all payment types

# To get the probability of A
A_credit = (orders_payment_reviews['payment_type_y'] == "credit_card").sum()
total_payment = len(orders_payment_reviews['payment_type_y'])
p_credit = round((A_credit / total_payment)*100, 4)
print(f'P(A) = {p_credit}')

# To get the probability of B
B = (orders_payment_reviews["review_score"]==5).sum()
total_review = len(orders_payment_reviews['review_score'])
p_B = round((B / total_review)*100, 4)
print(f'P(B) = {p_B}')

# To get P(review=5 | payment)
# Number of B | A
favourable_cond = ((orders_payment_reviews["review_score"]==5) & 
                   (orders_payment_reviews["payment_type_y"]=="credit_card")).sum()
total_cond = (orders_payment_reviews["review_score"]).sum()
prob_fav = round((favourable_cond / total_cond)*100, 4)

ans = (prob_fav*p_credit) / p_B
print(ans)

P(A) = 65.3898
P(B) = 57.0969
10.651901062929861


In [95]:
# For boleto payment;  A = Credit card; B = Review(5). B remains constant for all payment types

# To get the probability of A
A_credit = (orders_payment_reviews['payment_type_y'] == "boleto").sum()
total_payment = len(orders_payment_reviews['payment_type_y'])
p_credit = round((A_credit / total_payment)*100, 4)
print(f'P(A) = {p_credit}')

# To get the probability of B
B = (orders_payment_reviews["review_score"]==5).sum()
total_review = len(orders_payment_reviews['review_score'])
p_B = round((B / total_review)*100, 4)
print(f'P(B) = {p_B}')

# To get P(review=5 | payment)
# Number of B | A
favourable_cond = ((orders_payment_reviews["review_score"]==5) & 
                   (orders_payment_reviews["payment_type_y"]=="boleto")).sum()
total_cond = (orders_payment_reviews["review_score"]).sum()
prob_fav = round((favourable_cond / total_cond)*100, 4)

ans = (prob_fav*p_credit) / p_B
print(ans)

P(A) = 16.0897
P(B) = 57.0969
0.6398186915576853


In [96]:
# For boleto payment;  A = Credit card; B = Review(5). B remains constant for all payment types

# To get the probability of A
A_credit = (orders_payment_reviews['payment_type_y'] == "voucher").sum()
total_payment = len(orders_payment_reviews['payment_type_y'])
p_credit = round((A_credit / total_payment)*100, 4)
print(f'P(A) = {p_credit}')

# To get the probability of B
B = (orders_payment_reviews["review_score"]==5).sum()
total_review = len(orders_payment_reviews['review_score'])
p_B = round((B / total_review)*100, 4)
print(f'P(B) = {p_B}')

# To get P(review=5 | payment)
# Number of B | A
favourable_cond = ((orders_payment_reviews["review_score"]==5) & 
                   (orders_payment_reviews["payment_type_y"]=="voucher")).sum()
total_cond = (orders_payment_reviews["review_score"]).sum()
prob_fav = round((favourable_cond / total_cond)*100, 4)

ans = (prob_fav*p_credit) / p_B
print(ans)

P(A) = 17.2744
P(B) = 57.0969
0.7232648469531622


In [ ]:
# For boleto payment;  A = Credit card; B = Review(5). B remains constant for all payment types

# To get the probability of A
A_credit = (orders_payment_reviews['payment_type_y'] == "debit_card").sum()
total_payment = len(orders_payment_reviews['payment_type_y'])
p_credit = round((A_credit / total_payment)*100, 4)
print(f'P(A) = {p_credit}')

# To get the probability of B
B = (orders_payment_reviews["review_score"]==5).sum()
total_review = len(orders_payment_reviews['review_score'])
p_B = round((B / total_review)*100, 4)
print(f'P(B) = {p_B}')

# To get P(review=5 | payment)
# Number of B | A
favourable_cond = ((orders_payment_reviews["review_score"]==5) & 
                   (orders_payment_reviews["payment_type_y"]=="debit_card")).sum()
total_cond = (orders_payment_reviews["payment_type_y"]).value_counts()
prob_fav = round((favourable_cond / )*100, 4)

ans = (prob_fav*p_credit) / p_B
print(ans)

(orders_payment_reviews["payment_type_y"]).value_counts()

P(A) = 1.2429
P(B) = 57.0969
0.024911593449031384


payment_type_y
credit_card    80916
voucher        21376
boleto         19910
debit_card      1538
not_defined        3
Name: count, dtype: int64

In [90]:
(orders_payment_reviews["review_score"]==5).sum()

np.int64(70654)

In [91]:
orders_payment_reviews["review_score"].sum()

np.float64(499408.0)

In [75]:
orders_payment_reviews.head()

order_id                     customer_id_x  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
4  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status_x order_purchase_timestamp_x  order_approved_at_x  \
0      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
1      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
2      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
3      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
4      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date_x order_delivered_customer_date_x  \
0            2017-10-04 19:55:00             2017-10-10 21:25:13   
1            2017-10-04 19:55:00             2017-10-10 21:25:13   
2            2017-10-04 19:55:00             2017-10-10 21:25:13   
3            2017-10-04 19:55:00             2017-10-10 21:25:13   
4            2017-10-04 19:55:00             2017-10-10 21:25:13   

  order_estimated_delivery_date_x              customer_unique_id_x  \
0             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
4             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix_x  ... order_delivered_customer_date_y  \
0                      3149.0  ...             2017-10-10 21:25:13   
1                      3149.0  ...             2017-10-10 21:25:13   
2                      3149.0  ...             2017-10-10 21:25:13   
3                      3149.0  ...             2017-10-10 21:25:13   
4                      3149.0  ...             2017-10-10 21:25:13   

  order_estimated_delivery_date_y              customer_unique_id_y  \
0             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
2             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
3             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
4             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

  customer_zip_code_prefix_y  customer_city_y  customer_state_y  \
0                       3149        sao paulo                SP   
1                       3149        sao paulo                SP   
2                       3149        sao paulo                SP   
3                       3149        sao paulo                SP   
4                       3149        sao paulo                SP   

  payment_sequential_y  payment_type_y payment_installments_y payment_value_y  
0                  1.0     credit_card                    1.0           18.12  
1                  1.0     credit_card                    1.0           18.12  
2                  1.0     credit_card                    1.0           18.12  
3                  3.0         voucher                    1.0            2.00  
4                  3.0         voucher                    1.0            2.00  

[5 rows x 37 columns]

In [76]:
subset = orders_payment_reviews['payment_type_y']
subset.value_counts()

payment_type_y
credit_card    80916
voucher        21376
boleto         19910
debit_card      1538
not_defined        3
Name: count, dtype: int64

**Interpretation**

1. `P(Credit Card | 5 stars)` = Among all 5-star reviews, what fraction used credit card? Among 5-star reviews, `75.29%` used credit card

2. `P(Voucher | 5 stars)` = Among all 5-star reviews, what fraction used voucher? Among 5-star reviews, `5.11%` used boleto.

3. `P(Boleto | 5 stars)` = Among all 5-star reviews, what fraction used boleto? `4.52%` of 5-star reviews uses Boleto payment method.

Unlike P(5 | payment)- conditional probability , these will sum to approximately 100% across all payment methods, because every 5-star order must fall into exactly one payment category.

**Expectation**

This is what you would expect from an experiment if you repeat an experiment many times.

In practice, this means taking the average of the outcome of an experiment, the resulting value is the expectation. In statistical terms, we are talking about the mean(the centre of the distribution).

So when we talk about the baseline performance, we are actually talking about expectation. Examples of performance are average delivery time, average customer spend, average rating

We use it to find out;
 - How long a customer typically wait to get his or her delivery?
 - What is the expected revenue per order?
 - What is the overall satisfaction level?
 - What is the expected sale per seller?

**Variance**
- This explain how far the result of an experiment is from what is expected. When this value is very high, we would conclude that outcome of the experiment is not consistent, it is not predictable, it is unstable. When the value is low, it speaks of predictability, consistency, and stability. In summary, the wider the spread the higher the risk, the narrower the spread the lower the risk and reliability.
- Note that variance is more interpretable when it is converted to standard deviation(Same unit as your data)


**Usecase**
- The combination of these two are use for risk assessment
- They are also used for measuring reliability - how consistent are deliveries, ratings, or payments?
- They are also use to make data driven decision - To choose sellers, couriers, products with stable perfomance not just high averages.
- Uncertainty estimation (used in confidence intervals and hypothesis testing)
- Machine Learning (feature scaling and PCA rely on mean & variance)

- **Lets see some examples**
   - Mean delivery time = 10 days, SD = 2 days
     - Meaning that “Most customers get deliveries between 8–12 days. Service is predictable.”

   - Mean delivery time = 10 days, SD = 6 days
     - Meaning that “Some customers get deliveries in 4 days, others in 16. Unreliable service.”

   - Mean review score = 4.2, SD = 0.5
     - “Almost everyone gives between 4–5 stars. Strong and consistent reputation.”

   - Mean review score = 4.2, SD = 1.8
     - “Some give 5, others 1. The average hides mixed experiences.”

**Question1**

What is the average delivery time (expectation) and variance (spread/consistency) of deliveries?

In [114]:
orders_payment_reviews.shape

(123744, 38)

In [115]:
# First convert the datetime columns to pandas dataframe

# Convert to datetime
orders_payment_reviews['order_purchase_timestamp_x'] = pd.to_datetime(orders_payment_reviews['order_purchase_timestamp_x'])
orders_payment_reviews['order_delivered_customer_date_x'] = pd.to_datetime(orders_payment_reviews['order_delivered_customer_date_x'])

# Calculate delivery time in days
orders_payment_reviews['delivery_time_days'] = (orders_payment_reviews['order_delivered_customer_date_x'] - orders_payment_reviews['order_purchase_timestamp_x']).dt.days

# Drop missing values
delivery_times = orders_payment_reviews['delivery_time_days'].dropna()

# Get the expectation(mean)
mean_delivery = delivery_times.mean()

# Get the variance and standard deviation
variance_delivery = delivery_times.var()
std_delivery = delivery_times.std()


print('Expected delivery time(mean): ', round(mean_delivery, 2), 'days')
print('Variance of delivery time: ', round(variance_delivery, 2), 'days')
print('Standard Deviation of delivery time:', round(std_delivery, 2), 'days')


Expected delivery time(mean):  12.09 days
Variance of delivery time:  87.8 days
Standard Deviation of delivery time: 9.37 days


**Interpretation**
With the mean being 12 days and the standard deviation being 9 days, this means most customers get their deliveries between 3 and 21 days. This makes the service unreliable.

Another shoot in the leg for this company is their variance. A variance of 87.8 indicates a very wide spread, and as such, poses a high risk for the users.

In conclusion, given the mean standard deviation and variance, I infer that the service of this company is inconsistent, unpredictable, and unstable, making them highly unreliable.

**Question2**

What is the expected customer review score and how much do reviews vary?

In [116]:
mean_review = orders_payment_reviews['review_score'].mean()

#Variance and Standard Deviation
review_std = orders_payment_reviews['review_score'].std()
review_variance = orders_payment_reviews['review_score'].var()


print('Expected delivery time(mean): ', round(mean_review, 2), 'days')
print('Variance of delivery time: ', round(review_variance, 2), 'days')
print('Standard Deviation of delivery time:', round(review_std, 2), 'days')

Expected delivery time(mean):  4.06 days
Variance of delivery time:  1.88 days
Standard Deviation of delivery time: 1.37 days



**Interpretation**

The mean review score shows the overall satisfaction (e.g., ~4 stars = mostly happy customers). The std tells us that most customers give between 3 and 5 stars. If Std was very high (e.g., 2), it would mean customers are divided (some give 1, others give 5). But here, the service reputation is fairly consistent.

### **Sampling & Estimation**

Random sampling means selecting a subset of your data without bias so that it fairly represents the whole population.

In business terms, imagine you can’t survey all customers. Instead, you ask a random group and use their responses to estimate the overall trend.

- Lets see how we can work with this in our dataset

In [117]:
orders_payment_reviews.head(2)

order_id                     customer_id_x  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status_x order_purchase_timestamp_x  order_approved_at_x  \
0      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
1      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date_x order_delivered_customer_date_x  \
0            2017-10-04 19:55:00             2017-10-10 21:25:13   
1            2017-10-04 19:55:00             2017-10-10 21:25:13   

  order_estimated_delivery_date_x              customer_unique_id_x  \
0             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix_x  ... order_estimated_delivery_date_y  \
0                      3149.0  ...             2017-10-18 00:00:00   
1                      3149.0  ...             2017-10-18 00:00:00   

               customer_unique_id_y  customer_zip_code_prefix_y  \
0  7c396fd4830fd04220f754e42b4e5bff                        3149   
1  7c396fd4830fd04220f754e42b4e5bff                        3149   

  customer_city_y  customer_state_y  payment_sequential_y payment_type_y  \
0       sao paulo                SP                   1.0    credit_card   
1       sao paulo                SP                   1.0    credit_card   

   payment_installments_y payment_value_y delivery_time_days  
0                     1.0           18.12                8.0  
1                     1.0           18.12                8.0  

[2 rows x 38 columns]

In [120]:
# Population mean (true mean from entire dataset
pop_mean_payment = orders_payment_reviews['payment_value_y'].mean()

# Take a random sample of 50 orders
sample_50 = orders_payment_reviews['payment_value_y'].sample(50, random_state=42)
sample_mean_50 = sample_50.mean()

sample_200 = orders_payment_reviews['payment_value_y'].sample(200, random_state=42)
sample_mean200 = sample_200.mean()

print('Population mean payment value: ', round(pop_mean_payment, 2))
print('Sample mean (n= 50)', round(sample_mean_50, 2))
print('Sample population (n=200):', round(sample_mean200, 2))

Population mean payment value:  135.85
Sample mean (n= 50) 86.99
Sample population (n=200): 149.16


**Interpretation**

- Population mean = “the real average spending across all customers.”

- Sample mean (50 orders) may be off (underestimate or overestimate).

- Sample mean (200 orders) should be closer to the population mean.

This shows that why random sampling is powerful we can get a good estimate of the population without looking at every order.

### **Law of Large Numbers (LLN)**

You might wonder why big data is a reliable estimate, the idea behind it is the law of large numbers. Imagine you are asking people are the nature of the AI fellowship, the more people you ask, the more likely to build up your expectation or mindset about what the quality of the fellowship.

- The Law of Large Numbers says,

"As the sample size increases, the sample mean → converges to the true population mean"


In [126]:
import numpy as np

#Draw increasing sample sizes and track their names
sample_sizes = [10, 50, 100, 500, 1000]
sample_means = []

for n in sample_sizes:
    sample = orders_payment_reviews['payment_value_y'].sample(n, random_state=234)
    sample_means.append(sample.mean())

# sHOW RESULTS
for n, mean in zip(sample_sizes, sample_means):
    print(f'Sample size={n}: Sample size mean={round(mean, 2)}')

print('Population mean (true):', round(pop_mean_payment, 2))

Sample size=10: Sample size mean=116.29
Sample size=50: Sample size mean=105.42
Sample size=100: Sample size mean=98.99
Sample size=500: Sample size mean=144.51
Sample size=1000: Sample size mean=139.49
Population mean (true): 135.85


**Interpretation**

The result above should expplain the reliability of big data.

With 10 samples, the mean may jump around (unreliable).Ad with 1000+ samples the  mean is almost the same as population mean.

### **Central Limit Theorem (CLT)**

In [127]:
orders_payment_reviews.head(2)

order_id                     customer_id_x  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status_x order_purchase_timestamp_x  order_approved_at_x  \
0      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   
1      delivered        2017-10-02 10:56:33  2017-10-02 11:07:15   

  order_delivered_carrier_date_x order_delivered_customer_date_x  \
0            2017-10-04 19:55:00             2017-10-10 21:25:13   
1            2017-10-04 19:55:00             2017-10-10 21:25:13   

  order_estimated_delivery_date_x              customer_unique_id_x  \
0             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1             2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix_x  ... order_estimated_delivery_date_y  \
0                      3149.0  ...             2017-10-18 00:00:00   
1                      3149.0  ...             2017-10-18 00:00:00   

               customer_unique_id_y  customer_zip_code_prefix_y  \
0  7c396fd4830fd04220f754e42b4e5bff                        3149   
1  7c396fd4830fd04220f754e42b4e5bff                        3149   

  customer_city_y  customer_state_y  payment_sequential_y payment_type_y  \
0       sao paulo                SP                   1.0    credit_card   
1       sao paulo                SP                   1.0    credit_card   

   payment_installments_y payment_value_y delivery_time_days  
0                     1.0           18.12                8.0  
1                     1.0           18.12                8.0  

[2 rows x 38 columns]

In [128]:
orders_payment_reviews.columns

Index(['order_id', 'customer_id_x', 'order_status_x',
       'order_purchase_timestamp_x', 'order_approved_at_x',
       'order_delivered_carrier_date_x', 'order_delivered_customer_date_x',
       'order_estimated_delivery_date_x', 'customer_unique_id_x',
       'customer_zip_code_prefix_x', 'customer_city_x', 'customer_state_x',
       'payment_sequential_x', 'payment_type_x', 'payment_installments_x',
       'payment_value_x', 'review_id', 'review_score', 'review_comment_title',
       'review_comment_message', 'review_creation_date',
       'review_answer_timestamp', 'customer_id_y', 'order_status_y',
       'order_purchase_timestamp_y', 'order_approved_at_y',
       'order_delivered_carrier_date_y', 'order_delivered_customer_date_y',
       'order_estimated_delivery_date_y', 'customer_unique_id_y',
       'customer_zip_code_prefix_y', 'customer_city_y', 'customer_state_y',
       'payment_sequential_y', 'payment_type_y', 'payment_installments_y',
       'payment_value_y', 'delivery

In [ ]:
# lets collect 100 samples of review_scores, take the mean. Repeat the same process a thousand times.
# Then take all the 1000 means use it to plot a histogram and lets see what the shape or distribution would look like.
sample_means = []
for i in range(1000):           #repeat 1000 times
       sample = orders_payment_reviews['re']